### ISSUE TITLE
Performant bootstrap code

### ISSUE THREAD

**Issue by bgreenwell** [AFIT-R/surrogateResiduals#5]
I think this should be done last! If Rcpp is the answer, we'll need to put some thought into it, but issues #3 and #4 need to be worked out first. Also, the default number of bootstrap replicates should be 1.

The main obstacle I'm seeing with writing this bottleneck in C++ is the fact that the inner loop requires sampling from a truncated distirbution available in another package. You can pass R functions to Rcpp using typedef `Function`, but calling it inside a loop *will make the C++ code slower than the original R code!* This means we may have to define our own sampling function? But there are other ways to speed up the sampling without Rcpp that we can try first. A good place to look for ideas is the [`bootMer`](https://github.com/lme4/lme4/blob/master/R/bootMer.R) function from the well known lme4 package.

**Comment by bgreenwell** [comment]
Anyone know why we're using a multivariate function (`rtmvnorm`) to generate one surrogate at a time (we only need to sample from a univariate distribution)? This seems to be extremely inefficient (and slow). Here is some benchmark code comparing it to a faster approach:

```r
# Load required packages
library(ggplot2)
library(microbenchmark)
library(tmvtnorm)
library(truncdist)

# Slow loop using multivariate distribution: tmvtnorm
fun1 <- function() {
  res <- numeric(100)
  for (i in 1:100) {
    res[i] <- rtmvnorm(1, mean = 0, sigma = 1, lower = -1, upper = 1)
  }
}

# Faster loop using univariate distribution: truncdist
fun2 <- function() {
  res <- numeric(100)
  for (i in 1:100) {
    res[i] <- rtrunc(1, spec = "norm", a = -1, b = 1, mean = 0, sd = 1)
  }
}

# Compare timings
mb <- microbenchmark(fun1(), fun2(), times = 100)
autoplot(mb)
print(mb)  # using truncdist is about 10x faster!
```

**Comment by bradleyboehmke** [comment]
@bgreenwell  No.  This came straight from Dungang's code.  He probably wasn't thinking efficiency when he built it and initially we just ported the code over into a package.  Let's build the efficiency in as long as we don't lose the theoretical correctness of the approach.
